### PR TITLE
further improve 2d set_property function

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
*improved the function so that users do not need to delete and add back
entities in order to specify initial locations of entities.

### CODE REVIEW

File: games/xworld/maps/XWorldDialogMap.py [hunk 1]
```diff
@@ -12,8 +12,8 @@ def __init__(self, item_path, start_level=0):
 
     def _configure(self, select_class=True):
         self.set_dims(1, 1)
-        self.set_entity(type="agent")
-        self.set_entity(type="goal")
+        self.set_entity(type="agent", loc=(0, 0), force_occupy=True)
+        self.set_entity(type="goal", loc=(0, 0), force_occupy=True)
         if select_class:
             self.select_goal_classes() # re-select goal class for a new session
```
Comment (yu239) [inline games/xworld/maps/XWorldDialogMap.py]: this will have conflicts if later a maze is randomly generated.

Comment (Haichao-Zhang) [reply games/xworld/maps/XWorldDialogMap.py]: Now makes the maze_generation and optional choice and generates warning message for over-writing pre-specified locations if the maze generation is turned on.

File: games/xworld/maps/xworld_env.py [hunk 2]
```diff
@@ -193,17 +203,26 @@ def check_or_get_value(valid_value_set, is_continuous=False):
...
+                if entity.loc in self.available_grids:
+                    self.available_grids.remove(entity.loc)
             if property == "name":
                 entity.name = check_or_get_value(self.get_all_possible_names(entity.type))
                 # update id once name is changed
                 entity.id = "%s_%s" % (entity.name, self.entity_nums[entity.type])
-            if property == "asset_path" or property == "name":
-                # update the asset_path once the name is changed
+                entity.asset_path = None  # update the asset_path once the name is changed
+            if property == "asset_path":
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: If the user sets asset_path, we should also automatically set name and id for it.

File: games/xworld/maps/xworld_env.py [hunk 3]
```diff
@@ -409,30 +428,36 @@ def __instantiate_entities(self):
...
+                    e.loc = None # remove the pre-set location when maze_generation is on
+                self.set_property(e)
+            ## add back some empty grids
+            self.available_grids += blocks[len(self.get_blocks()):]
+        else:
+            ## instantiate properties for each entity
+            for i, e in enumerate(self.entities):
+                self.set_property(e)
+
+        if self.maze_generation:
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: merge this code block into the previous branch

File: games/xworld/maps/xworld_env.py [hunk 4]
```diff
@@ -409,30 +428,36 @@ def __instantiate_entities(self):
...
+            ## first remove all maze blocks from the available set
+            for b in blocks:
+                if b in self.available_grids:
+                    self.available_grids.remove(b)
+            ## instantiate properties for each entity
+            for i, e in enumerate(self.entities):
+                if e.loc is not None:
+                    warnings.warn("Maze generation is on! Over-writing pre-specified location %s!" % (e.loc,))
+                    e.loc = None # remove the pre-set location when maze_generation is on
+                self.set_property(e)
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: what if the entity type is "block"? Originally my code will only set loc if it's not a block.

File: games/xworld/maps/xworld_env.py [hunk 5]
```diff
@@ -193,17 +203,26 @@ def check_or_get_value(valid_value_set, is_continuous=False):
             assert property in entity.__dict__.keys() and property in default_dict.keys(), \
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Why call it "context_dict"?

File: games/xworld/maps/xworld_env.py [hunk 6]
```diff
@@ -193,17 +203,26 @@ def check_or_get_value(valid_value_set, is_continuous=False):
...
                 "invalid property name: %s is provided" % property
             value = context_dict[property]
-            if value == False: # should only be used for properties with values
+            if value is False: # should only be used for properties with values
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Why call it "context_dict"?

File: games/xworld/maps/xworld_env.py [hunk 7]
```diff
@@ -193,17 +203,26 @@ def check_or_get_value(valid_value_set, is_continuous=False):
...
                 assert entity.__dict__[property], "no existing value for property %s" % property
                 continue
+            elif value is None and entity.__dict__[property] is None:
+                #reinstantiate if property value doesn't exist
+                value = True
+            elif value is None:
+                continue
```
Comment (yu239) [inline games/xworld/maps/xworld_env.py]: Why do you need True, False, and None for value? The logic here is complex. I believe a cleaner design exists.

File: games/xworld/maps/xworld_env.py (hunk 2 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: Code revised to handle this case.

File: games/xworld/maps/xworld_env.py (hunk 3 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: merged

File: games/xworld/maps/xworld_env.py (hunk 4 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: Changed the code. Now also skip setting the loc for blocks.

File: games/xworld/maps/xworld_env.py (hunk 6 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: name changed.

File: games/xworld/maps/xworld_env.py (hunk 7 above)
Comment (Haichao-Zhang) [reply games/xworld/maps/xworld_env.py]: These options  are included for handling existing use cases. May re-design in the future.

**Pull request merged**